### DISCUSSION TITLE
How do I specify `arm64e` vs. `x86` in dependency fetching?

### DISCUSSION THREAD

**Discussion by femtomc** [python-poetry/poetry#6574] [Q&A]
I'm running into issues where `poetry` gets the "most relaxed" `none` platform wheels -- but I'm running on Apple M1 and so I can errors related to having a mix of `arm64` wheels and `x86` wheels, even for packages *where I know a native ARM wheel is available*.

I'd really like to be able to specify: get all platform consistent wheels.

How do I do this?

**Comment by neersighted** [comment]
Poetry should be preferring arm wheels wherever available -- we'll need your pyproject.toml and lock file to better understand what is going on.

**Reply by femtomc** [reply]
Cool! Here's my `pyproject.toml`

```
[tool.poetry]
... (occluded for private dev reasons)

[tool.poetry.dependencies]
python = ">=3.10, <3.11"
jax = "^0.3.15"
jaxlib = "^0.3.15"
numpy = "^1.23.2"

[tool.poetry.group.dev.dependencies]
matplotlib = "^3.5.3"
seaborn = "^0.11.2"
optax = "^0.1.3"
black = "^22.6.0"
pytest = "^7.1.2"
coverage = "^6.4.3"
flake8 = "^5.0.4"
flake8-bugbear = "^22.7.1"
mypy = "^0.971"
pre-commit = "^2.20.0"
pytest-benchmark = "^3.4.1"
autoflake8 = "^0.3.2"
Sphinx = "^5.1.1"
jupyter-sphinx = "^0.4.0"
sphinx-material = "^0.0.35"


[tool.black]
line-length = 80
extend-exclude = '''
/(
  experiments
)/
'''

[tool.coverage.run]
omit = [".*", "*/site-packages/*"]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

The lockfile is rather big, what's the best way to communicate that?

**Reply by femtomc** [reply]
Re -- the packages I'm having issue with are dependencies of `jupyter-sphinx` -- packages like `psutil` and `lxml` which provide ARM wheels, but somehow get x86 wheels when I use `poetry install`.

When I try to build my documentation, I run into issues like 
```
Extension error:
Could not import extension jupyter_sphinx (exception: dlopen(/Users/mccoybecker/opt/miniconda3/envs/jax/lib/python3.10/site-packages/psutil/_psutil_osx.cpython-310-darwin.so, 0x0002): tried: '/Users/mccoybecker/opt/miniconda3/envs/jax/lib/python3.10/site-packages/psutil/_psutil_osx.cpython-310-darwin.so' (mach-o file, but is an incompatible architecture (have (x86_64), need (arm64e))))
```

inside my poetry environment.

E.g.

```
poetry run make -C docs html
Running Sphinx v5.1.1

Extension error:
Could not import extension jupyter_sphinx (exception: dlopen(/Users/mccoybecker/opt/miniconda3/envs/jax/lib/python3.10/site-packages/psutil/_psutil_osx.cpython-310-darwin.so, 0x0002): tried: '/Users/mccoybecker/opt/miniconda3/envs/jax/lib/python3.10/site-packages/psutil/_psutil_osx.cpython-310-darwin.so' (mach-o file, but is an incompatible architecture (have (x86_64), need (arm64e))))
make: *** [html] Error 2
```

**Reply by neersighted** [reply]
Try Gisting the lockfile -- it's going to be important to understand exactly what's going on

**Reply by femtomc** [reply]
Here's the lockfile: https://gist.github.com/femtomc/8483205347c3ebb9fdf4637f2d95a976

**Reply by neersighted** [reply]
https://gist.github.com/femtomc/8483205347c3ebb9fdf4637f2d95a976#file-poetry-lock-L2559-L2581
https://gist.github.com/femtomc/8483205347c3ebb9fdf4637f2d95a976#file-poetry-lock-L2178-L2249

Right of the bat I'm not seeing any macOS ARM wheels available -- I'd suggest you first `poetry update` and check if you get any eligible wheels. If not, the next step would be adding explicit constraints to the problematic packages that pin them to versions with macOS ARM wheels, and then running the solver to see if it can come up with a solution with those versions forced.

I am curious though -- pip/packaging are what in theory could allow for the usage of amd64 wheels, as Poetry just uses packaging to gather facts from the target environment. I'm not sure how incompatible wheels are getting sucked in given pip **inside** the target environment should choke and die on an incompatible wheel (Poetry's logic is still double-checked by pip during package install).

**Reply by femtomc** [reply]
I don't fully understand this either -- I can roughly setup a working environment by removing these packages manually and installing them via `conda` -- or some variant of this manual manipulation. So it seems like there is a path that works, but not with `pip`.

**Reply by neersighted** [reply]
Note that when you install with `conda`/`pip` manually you are ignoring version resolution concerns -- Poetry will only install versions from the lock file, and only versions that satisfy constraints. The reason you're not getting ARM wheels is because Poetry has either selected a version that does not have ARM wheels due to constraints, or you need to `poetry update` and regenerate your lockfile since new versions have become available.

Please try that and let me know if you manage to lock a version with ARM wheels.

**Reply by femtomc** [reply]
`poetry update` provided no new versions. If I manually remove and try with `conda`, I get this:
```
conda install lxml
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/mccoybecker/opt/miniconda3/envs/jax

  added / updated specs:
    - lxml


The following NEW packages will be INSTALLED:

  icu                pkgs/main/osx-arm64::icu-68.1-hc377ac9_0
  libiconv           pkgs/main/osx-arm64::libiconv-1.16-h1a28f6b_2
  libxml2            pkgs/main/osx-arm64::libxml2-2.9.14-h8c5e841_0
  libxslt            pkgs/main/osx-arm64::libxslt-1.1.35-h9833966_0
  lxml               pkgs/main/osx-arm64::lxml-4.9.1-py310h2fae87d_0


Proceed ([y]/n)?
```
which seems to imply that *there are wheels available for that version*

**Reply by femtomc** [reply]
Or maybe I'm misunderstanding.

**Reply by neersighted** [reply]
Conda is not the same as any other tool -- Conda provides their own binaries and wheels that are built by them, and not by the maintainers as in the regular PyPI ecosystem.

You should check individual releases on PyPI to see what packages are available -- indeed, I did that for lxml 4.9.1 (the latest) and there are no ARM wheels: https://pypi.org/project/lxml/#files

**Reply by femtomc** [reply]
Ah got it -- right.

**Reply by femtomc** [reply]
So basically, I'm SOL if PyPI doesn't provide an ARM wheel?

**Reply by femtomc** [reply]
Which seems to be the case for `lxml` -- as I continue to search versions, not having luck.

**Reply by neersighted** [reply]
You could add direct URL dependencies (with correct platform markers specified by yourself) on the Conda wheels if you need to (as I don't think they provide a PEP 508-compatible repository Poetry can resolve from). I'd also suggest opening an issue to see if we can figure out how Poetry (or rather, packaging) is selecting a x86_64 wheel on an ARM Mac (and why pip installs it) -- though I would appreciate you testing with a non-Conda Python version first.

I wonder if the Python that Conda builds pretends to be amd64 compatible (or if that's something Python does in general and I've never noticed -- my knowledge here is limited) as Rosetta 2 can load an Intel shared object into an ARM64 process. Attempting to reproduce the incorrect wheels being installed on a clean Python.org build of Python would be appreciated. If it does reproduce on a pyenv/python-build produced version, I would most certainly open an issue.

**Reply by neersighted** [reply]
Also, if we make sure to not select incorrect wheels, a sdist build should work just fine. Wheels are just a way to avoid compiling native code (or needing the native libraries/headers locally), but Poetry can build lxml from source just fine... So understanding why packaging is selecting the wrong wheels here is key, as if we fix that, the sdist-based fallback will work.

**Reply by femtomc** [reply]
Okay, I have to defer attempting to solve this until after some other work -- but I'll return to it later this week and add some more information.

**Reply by neersighted** [reply]
One final thought -- since you're on 1.2.x you can take advantage of `installer.no-binary` to force Poetry to ignore wheels -- `poetry config --local installer.no-binary lxml` should force a sdist-based install of lxml, for example.
